core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -485,7 +520,18 @@ class Partition(val topic: String,
     laggingReplicas
   }
 
-  def appendRecordsToLeader(records: MemoryRecords, isFromClient: Boolean, requiredAcks: Int = 0) = {
+  def appendRecordsToFollower(records: MemoryRecords, isFuture: Boolean) {
+    // The read lock is needed to avoid race condition while ReplicaAlterDirThread is executing
+    // maybeDeleteAndSwapFutureReplica() to replace follower replica with the future replica
+    inReadLock(leaderIsrUpdateLock) {
+      if (isFuture)
+        getReplica(Request.FutureLocalReplicaId).get.log.get.appendAsFollower(records)
```

Comment:
Should we check whether  getReplica(Request.FutureLocalReplicaId).get.log is None (in case we get read lock immediately after write lock is released) ?
See code at line 197 above.

Reply:
@tedyu maybeReplaceCurrentWithFutureReplica() will not cause NoSuchElementException here because there the thread that calls maybeReplaceCurrentWithFutureReplica() for a given partition is guaranteed to be the same ReplicaAlterDirThread that appends record to the future replica of this partition.